File: app/src/main/java/com/example/TibiaTools/View/HouseActivity.java
package com.example.TibiaTools.View;

import android.content.Intent;
import android.os.Build;
import android.os.Bundle;

import com.example.TibiaTools.View.ViewModel.ViewModelHouses;
import com.example.TibiaTools.recyclerview.Adapters.AdapterRecyclerViewHouses;
import com.example.TibiaTools.recyclerview.ItemsRecyclerViewHouses;
import com.example.TibiaTools.utilidades.DataHighScores;
import com.example.TibiaTools.utilidades.Spinners;
import com.example.ttools.R;
import com.example.ttools.databinding.ActivityHouseBinding;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Handler;
import android.os.Looper;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Toast;

import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;



import java.util.ArrayList;
import java.util.Objects;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class HouseActivity extends AppCompatActivity implements AdapterView.OnItemClickListener {

    private ActivityHouseBinding binding;
    String mundo;
    DataHighScores dataHighScores = new DataHighScores();
    AutoCompleteTextView spinnerWorlds, spinnerCitys;
    ArrayAdapter<String> adapterWorlds,adapterCitys;
    //Recyclerview
    RecyclerView recyclerView;
    AdapterRecyclerViewHouses adapterRecyclerViewHouses;
    ArrayList<ItemsRecyclerViewHouses> list_houses = new ArrayList<>();
    ViewModelProvider viewModelProvider;
    ViewModelHouses viewModelHouses;

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityHouseBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        setSupportActionBar(binding.toolbar);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        recyclerView = findViewById(R.id.recycler_houses);
        recyclerView.setHasFixedSize(true);
        linearLayoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        recyclerView.setLayoutManager(linearLayoutManager);
        adapterRecyclerViewHouses = new AdapterRecyclerViewHouses(list_houses);
        recyclerView.setAdapter(adapterRecyclerViewHouses);
        Objects.requireNonNull(getSupportActionBar()).setDisplayHomeAsUpEnabled(true); //Aparicion del boton regresar en el action bar
        binding.getRoot().findViewById(R.id.carga_houses).setVisibility(View.VISIBLE);
        viewModelProvider = new ViewModelProvider(this);
        viewModelHouses = viewModelProvider.get(ViewModelHouses.class);
        spinnerWorlds = findViewById(R.id.spinner_mundos);
        spinnerCitys = findViewById(R.id.spinner_citys);
        spinnerWorlds.setOnItemClickListener(this);
        spinnerCitys.setOnItemClickListener(this);
        Threads();
        viewModelHouses.Worlds().observe(this, worlds -> {
            if (worlds != null){
                adapterWorlds = new ArrayAdapter<>(getApplicationContext(), R.layout.auto_complete, worlds);
                spinnerWorlds.setAdapter(adapterWorlds);
                binding.getRoot().findViewById(R.id.carga_houses).setVisibility(View.GONE);
            } else{
                Toast.makeText(getApplicationContext(),"Fallo en la opcion mundos.. Intente mas tarde", Toast.LENGTH_SHORT).show();
                binding.getRoot().findViewById(R.id.carga_houses).setVisibility(View.GONE);
            }
        });
        viewModelHouses.houses().observe(this, houses -> {
            list_houses.clear();
            if (houses != null){
                mundo = houses.getWorld();
                houses.getHouse_list().forEach(houseList -> {
                    String rented;
                    if (houseList.isRented()){
                        rented = "Ocupada";
                    } else{
                        rented = "Desocupada";
                    }
                    list_houses.add(new ItemsRecyclerViewHouses(
                            houseList.getName(),
                            String.valueOf(houseList.getSize()),
                            String.valueOf(houseList.getRent()),
                            rented,
                            String.valueOf(houseList.getHouse_id())
                    ));
                });
                //validar que el array de guildhall es diferente de null
                if (houses.getGuildhall_list()!= null) {
                    houses.getGuildhall_list().forEach(guildhallList -> {
                        String rented;
                        if (guildhallList.isRented()){
                            rented = "Ocupada";
                        } else{
                            rented = "Desocupada";
                        }
                        list_houses.add(new ItemsRecyclerViewHouses(
                                guildhallList.getName(),
                                String.valueOf(guildhallList.getSize()),
                                String.valueOf(guildhallList.getRent()),
                                rented,
                                String.valueOf(guildhallList.getHouse_id())
                        ));
                    });
                }
            }
            binding.getRoot().findViewById(R.id.carga_houses).setVisibility(View.GONE);
            adapterRecyclerViewHouses.notifyDataSetChanged();
        });

        adapterRecyclerViewHouses.setOnClickListener(view -> {
            //Enviando el id de la casa a al activity HousesInformation
            String id_house = list_houses.get(
                    recyclerView.getChildAdapterPosition(view)
            ).getHouseId();
            Intent intent = new Intent(HouseActivity.this, HousesInformation.class);
            intent.putExtra("ID",id_house);
            intent.putExtra("mundo", mundo);
            startActivity(intent);
        });

    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == android.R.id.home) { //aqui daremos el evento al boton regresar de nuestro action bar, haciendo uso de los ids del sistema android
            finish();// finalizamos la actividad
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void spinnersCity(){
        Spinners spinners = new Spinners();
        adapterCitys = new ArrayAdapter<>(getApplicationContext(), R.layout.auto_complete, spinners.LeerDataCitys(getResources().openRawResource(R.raw.data)));
        new Handler(Looper.getMainLooper()).post(()->{
            spinnerCitys.setAdapter(adapterCitys);
        });
    }

    //llamadas asincornas con hilos
    public void Threads(){
        Executor executor = Executors.newFixedThreadPool(1);
        executor.execute(this::spinnersCity);
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        if (!spinnerCitys.getText().toString().isEmpty() && !spinnerWorlds.getText().toString().isEmpty()){
            if (!spinnerCitys.getText().toString().equalsIgnoreCase("Seleccione una ciudad") && !spinnerWorlds.getText().toString().equalsIgnoreCase("Seleccione")){
                String mundo, ciudad;
                mundo = spinnerWorlds.getText().toString();
                ciudad = spinnerCitys.getText().toString();
                dataHighScores.setMundo(mundo);
                dataHighScores.setCiudad(ciudad);
                viewModelHouses.setHouses(dataHighScores.getMundo(),dataHighScores.getCiudad());
            }
            else {
                list_houses.clear();
                adapterRecyclerViewHouses.notifyDataSetChanged();
            }
        }
    }
}

File: app/src/main/java/com/example/TibiaTools/View/HouseActivity.kt
package com.example.TibiaTools.View

import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.TibiaTools.data.model.*
import com.example.TibiaTools.View.ViewModel.ViewModelHouses
import com.example.TibiaTools.recyclerview.Adapters.AdapterRecyclerViewHouses
import com.example.TibiaTools.recyclerview.ItemsRecyclerViewHouses
import com.example.TibiaTools.utilidades.DataHighScores
import com.example.TibiaTools.utilidades.Spinners
import com.example.ttools.R
import com.example.ttools.databinding.ActivityHouseBinding
import java.util.ArrayList
import java.util.Objects
import java.util.concurrent.Executors

class HouseActivity : AppCompatActivity(), AdapterView.OnItemClickListener {
    private lateinit var binding: ActivityHouseBinding
    private var mundo: String? = null
    private val dataHighScores = DataHighScores()
    private lateinit var spinnerWorlds: AutoCompleteTextView
    private lateinit var spinnerCitys: AutoCompleteTextView
    private var adapterWorlds: ArrayAdapter<String>? = null
    private var adapterCitys: ArrayAdapter<String>? = null
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapterRecyclerViewHouses: AdapterRecyclerViewHouses
    private val listHouses = ArrayList<ItemsRecyclerViewHouses>()
    private lateinit var viewModelProvider: ViewModelProvider
    private lateinit var viewModelHouses: ViewModelHouses

    @RequiresApi(Build.VERSION_CODES.N)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHouseBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)

        val linearLayoutManager = LinearLayoutManager(this).apply {
            orientation = LinearLayoutManager.VERTICAL
        }
        recyclerView = findViewById(R.id.recycler_houses)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = linearLayoutManager
        adapterRecyclerViewHouses = AdapterRecyclerViewHouses(listHouses)
        recyclerView.adapter = adapterRecyclerViewHouses

        Objects.requireNonNull(supportActionBar)?.setDisplayHomeAsUpEnabled(true)
        binding.root.findViewById<View>(R.id.carga_houses)?.visibility = View.VISIBLE

        viewModelProvider = ViewModelProvider(this)
        viewModelHouses = viewModelProvider[ViewModelHouses::class.java]

        spinnerWorlds = findViewById(R.id.spinner_mundos)
        spinnerCitys = findViewById(R.id.spinner_citys)
        spinnerWorlds.onItemClickListener = this
        spinnerCitys.onItemClickListener = this

        threads()

        viewModelHouses.Worlds().observe(this) { worlds ->
            if (worlds != null) {
                adapterWorlds = ArrayAdapter(applicationContext, R.layout.auto_complete, worlds)
                spinnerWorlds.setAdapter(adapterWorlds)
                binding.root.findViewById<View>(R.id.carga_houses)?.visibility = View.GONE
            } else {
                Toast.makeText(applicationContext, "Fallo en la opcion mundos.. Intente mas tarde", Toast.LENGTH_SHORT).show()
                binding.root.findViewById<View>(R.id.carga_houses)?.visibility = View.GONE
            }
        }

        viewModelHouses.houses().observe(this) { houses ->
            listHouses.clear()
            if (houses != null) {
                mundo = houses.world
                houses.house_list.forEach { houseList ->
                    val rented = if (houseList.rented) "Ocupada" else "Desocupada"
                    listHouses.add(
                        ItemsRecyclerViewHouses(
                            houseList.name,
                            houseList.size.toString(),
                            houseList.rent.toString(),
                            rented,
                            houseList.house_id.toString()
                        )
                    )
                }
                houses.guildhall_list?.forEach { guildhallList ->
                    val rented = if (guildhallList.rented) "Ocupada" else "Desocupada"
                    listHouses.add(
                        ItemsRecyclerViewHouses(
                            guildhallList.name,
                            guildhallList.size.toString(),
                            guildhallList.rent.toString(),
                            rented,
                            guildhallList.house_id.toString()
                        )
                    )
                }
            }
            binding.root.findViewById<View>(R.id.carga_houses)?.visibility = View.GONE
            adapterRecyclerViewHouses.notifyDataSetChanged()
        }

        adapterRecyclerViewHouses.setOnClickListener { view ->
            val position = recyclerView.getChildAdapterPosition(view)
            if (position != RecyclerView.NO_POSITION) {
                val idHouse = listHouses[position].houseId
                val name = listHouses[position].houseName
                val intent = Intent(this, HousesInformation::class.java).apply {
                    putExtra("ID", idHouse)
                    putExtra("mundo", mundo)
                    putExtra("name", name)
                }
                startActivity(intent)
            }
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun spinnersCity() {
        val spinners = Spinners()
        adapterCitys = ArrayAdapter(applicationContext, R.layout.auto_complete, spinners.LeerDataCitys(resources.openRawResource(R.raw.data)))
        Handler(Looper.getMainLooper()).post {
            spinnerCitys.setAdapter(adapterCitys)
        }
    }

    private fun threads() {
        val executor = Executors.newFixedThreadPool(1)
        executor.execute { spinnersCity() }
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        val cityText = spinnerCitys.text.toString()
        val worldText = spinnerWorlds.text.toString()
        if (cityText.isNotEmpty() && worldText.isNotEmpty()) {
            if (!cityText.equals("Seleccione una ciudad", ignoreCase = true) && !worldText.equals("Seleccione", ignoreCase = true)) {
                mundo = worldText
                val ciudad = cityText
                dataHighScores.mundo = mundo
                dataHighScores.ciudad = ciudad
                viewModelHouses.setHouses(dataHighScores.mundo, dataHighScores.ciudad)
            } else {
                listHouses.clear()
                adapterRecyclerViewHouses.notifyDataSetChanged()
            }
        }
    }
}
